### PR TITLE
Clean up main

### PR DESCRIPTION
**Pull request by psiborg112**
Main has accumulated a lot of VirtualStudio specific code which, combined with all of the ifdefs being used, has made it a little hard to get a quick overview of what is happening at launch. This PR encapsulates a lot of the deeplink and instance checking code in its own class.

Couple of notes: I've left the author field in vsInit.h and vsInit.cpp blank because it's not really my code. If anyone knows who originally authored it, and could add their name there, that would be great. Also, is the code that checks for an existing instance of the program really only meant to be run under windows? And as always, if someone could test that deeplinks still work as expected, that would be great.

### CODE REVIEW

File: src/main.cpp
Comment (ntonnaett) [inline src/main.cpp:289]: This instance doesn't seem to be used but just reset for Windows. Could be removed here and declared later when used.

**Comment by mattahorton** [comment]
> Also, is the code that checks for an existing instance of the program really only meant to be run under windows? 

Correct.

**Comment by psiborg112** [comment]
@mattahorton thanks for confirming that

File: src/gui/vsInit.cpp
Comment (dyfer) [inline src/gui/vsInit.cpp:87]: VS GUI [requires Qt 5.15](https://github.com/jacktrip/jacktrip/issues/852) anyway, AFAIR, so this could be possibly removed, I think.

**Comment by mattahorton** [comment]
I've just tested this. It seems that, on Windows, now the existing instance of JackTrip _also_ closes when the 2nd instance is run. 

The expected behavior is that the original instance continues running and the new one is prevented.

**Comment by psiborg112** [comment]
@mattahorton does it close or does it crash? Is there any debug output?

**Comment by psiborg112** [comment]
@OConnorStP This is purely a virtualstudio thing. I'm just trying to clean up the code so that there isn't VS stuff where there really shouldn't be.

**Comment by mattahorton** [comment]
> @mattahorton does it close or does it crash? Is there any debug output?

No crash output. No extra output at all. It just closes.

**Comment by psiborg112** [comment]
So what's the history of this code? Who authored it initially? (And can they take a look at it now?) What's the rationale behind it being only a windows thing?

**Comment by psiborg112** [comment]
OK, so it definitely crashes and I suspect the nested lambda is the issue. I think that by the time the internal lambda is called it has gone out of scope and no longer exists. I might have to actually derive vsInit from QObject so that these lambdas can be redefined as slots. (Which, to be honest, would also have the added benefit of making the code easier to follow.)

**Comment by psiborg112** [comment]
As an aside, I've tested the original code, and one undesired outcome is that if you're in classic mode then the second instance will launch a VS login screen if you're not already logged in. Which it really shouldn't do.

**Comment by psiborg112** [comment]
So that should fix it. (There was a null pointer issue because vs wasn't set in main when this was called.) This is some much needed cleanup, and should definitely still be merged once approved. But, having said that, neither this code nor the code that it replaces is fit for purpose and it needs to be rethought. It doesn't even remotely take into account the fact that VirtualStudio mode is not the only mode the program operates in. You could be in the middle of a classic mode session, connected to a server, and this code will still bring up the VS window. There's no reason for the user to assume that this is the same instance of the program that they're actively using, and if the VS login window gets stuck (as it did most of the time when I was testing the impacts of this) and the user closes it to get back to what they were doing, they end up closing the entire program.

This needs to be fixed, and it should not be on the open source developers to fix it. I suspect that a lot of the testing of new features by the VirtualStudio team is being done entirely within VS mode. Which is fine for things that don't impact the rest of the project. (And the architecture is modular enough that this will often be the case.) But it's unfair when your changes put an undue maintenance burden on the open source team. (Many of whom are time poor already.) This discovery already comes off the back of us having a 1.8.0 beta release that does not work at all when using JACK, and that doesn't work as expected with the RtAudio backend outside of VS mode. No one is expecting you to contribute improvements back to the standard interface or the command line. But please don't break these for us.

**Comment by psiborg112** [comment]
Rant aside, this latest commit should fix the issue that I've identified. (I'm on the road at the moment and haven't had a chance to actually test it yet, so absolutely do not merge without testing.)

**Comment by mattahorton** [comment]
I've been thinking a lot about regression testing for JackTrip and what that would take. I know nothing about the frameworks for that in Qt/C++. But it's clear there is a need.

**Review by mattahorton — APPROVED** [review]

**Comment by psiborg112** [comment]
It would be great to have automated tests, but even those wouldn't really have been necessary with #904. All it would have taken would have been to run the far reaching code (that touched far more than just VS mode and RtAudio) in any other configuration - the bugs happened instantly and 100% of the time. (Could not ask for a more reproducible bug.) I think it's fair for code reviewers to make the assumption that code has actually been run once it reaches the pull request stage, so that their focus is on more technical issues. And if it hasn't for some reason then the author needs to clearly state any limitations to their testing. (It's why I always stress where I haven't been able to test VS features, like deeplinks, where code I write butts up against their implementation.) I think there needs to be an understanding that if your code is outside of specific GUI territory and reaches into common areas then it absolutely needs to be tested in other modes, even if that means you need to collaborate with someone who can do the tests you can't.

**Pull request merged**
